crates/polars-pipe/src/executors/sinks/output/file_sink.rs

```diff
@@ -40,13 +44,18 @@ pub(super) fn init_writer_thread(
                     chunks.sort_by_key(|chunk| chunk.chunk_index);
                 }
 
-                for chunk in chunks.iter() {
-                    writer._write_batch(&chunk.data).unwrap()
+                for chunk in chunks.drain(0..) {
```

Comment:
Very crude solution. Can you test how this will perform with a single threaded `rechunk`?

Reply:
The issue is that if you just do a rechunk of `chunks`, you're limited to merging by the number of cores. E.g. with 12 cores, you merge 12 chunks, and that's all you merge. If you have 2 cores, you only merge 2 chunks, so the performance ends up being much worse if you have a stream of many small chunks. I can do another round of benchmarks tomorrow demonstrating that, but that's my memory from previous testing that led me to this strategy.

Reply:
Notice that `chunks` size is `morsels_per_sink` which is the size of the thread pool, basically. So the fewer cores you have, the less merging only `chunks` helps.

Reply:
> f you have 2 cores, you only merge 2 chunks, so the performance ends up being much worse if you have a stream of many small chunks. I can do another round of benchmarks tomorrow demonstrating that, but that's my memory from previous testing that led me to this strategy.

I don't really follow, because a `vstack_mut` is essentially the same. It only has one downside that is that we cannot know the size to allocate. So we can trigger multiple reallocations. With a `rechunk` we first do a pass to determine the required output memory size and then we trigger the memcopies. That's the reason I think we should just `rechunk`.

Reply:
So I took this as an opportunity to learn Divan, it's pretty cool: https://github.com/itamarst/benchmarks/blob/main/polars-vstack/benches/vstack.rs#L13

Running this I get:

```
Timer precision: 19 ns
vstack        fastest       │ slowest       │ median        │ mean          │ samples │ iters
╰─ vstack                   │               │               │               │         │
   ├─ 10000   885.2 µs      │ 1.049 ms      │ 921.5 µs      │ 925.5 µs      │ 100     │ 100
   │          90.36 Gitem/s │ 76.22 Gitem/s │ 86.8 Gitem/s  │ 86.43 Gitem/s │         │
   │          alloc:        │               │               │               │         │
   │            10009       │ 10009         │ 10009         │ 10009         │         │
   │            1.216 MB    │ 1.216 MB      │ 1.216 MB      │ 1.216 MB      │         │
   │          dealloc:      │               │               │               │         │
   │            2           │ 2             │ 2             │ 2             │         │
   │            8.084 KB    │ 8.084 KB      │ 8.084 KB      │ 8.084 KB      │         │
   │          grow:         │               │               │               │         │
   │            13          │ 13            │ 13            │ 13            │         │
   │            262.1 KB    │ 262.1 KB      │ 262.1 KB      │ 262.1 KB      │         │
   ╰─ 100000  9.565 ms      │ 9.966 ms      │ 9.696 ms      │ 9.713 ms      │ 100     │ 100
              83.63 Gitem/s │ 80.26 Gitem/s │ 82.5 Gitem/s  │ 82.35 Gitem/s │         │
              alloc:        │               │               │               │         │
                100009      │ 100009        │ 100009        │ 100009        │         │
                12.01 MB    │ 12.01 MB      │ 12.01 MB      │ 12.01 MB      │         │
              dealloc:      │               │               │               │         │
                2           │ 2             │ 2             │ 2             │         │
                8.084 KB    │ 8.084 KB      │ 8.084 KB      │ 8.084 KB      │         │
              grow:         │               │               │               │         │
                16          │ 16            │ 16            │ 16            │         │
                2.097 MB    │ 2.097 MB      │ 2.097 MB      │ 2.097 MB      │         │
```

I'm really not getting the impression that repeated `vstack_mut` leads to quadratic slowdown. Runtime is linear, allocated memory is linear and much smaller than underlying data: 100_000 * 1000 * 8 bytes is 800MB, but only 12MB gets allocated.

Next going to read the code for `vstack_mut` so I understand implementation better.

Reply:
So AFAICT `vstack_mut` eventually devolves to `new_chunks()` in `chunked_array/ops/append.rs`, which clones all the chunks individually, i.e. does not make them contiguous.

Reply:
I think we might be talking past each other here? So let me try again, and you can point out what if anything I got wrong.

Basically this PR is structured the way it is because a separate struct is necessary to accumulate state across batches.

The way `init_writer_thread` currently works:

1. It reads from a queue and adds `DataChunk` to `chunks`, a vec.
2. After receiving N `DataChunk`, where N is basically "the number of threads in the thread pool" it optionally sorts them, then writes the batch one by one to the underlying writer.

We know the individual chunks are too small, and we don't want to write them one by one when they're small, cause that's slow. So we want to combine them, somehow. There are two strategies one could take:

1. **Combine the batch:** Combine everything in `chunks` into a single DataFrame. This means we can combine at most N chunks at a time, no more, and for smaller thread pool sizes this might only be 2 chunks!
2. **Combine across multiple batches if necessary.** This is what this PR does. When you call `SemicontiguousVstacker.add()`, it might return nothing and keep the data in its internal buffer! So it can continue accumulating across write batches until it hits some limit (in current state, 4MB of data). For a 2.5KB-sized chunk, that means it's accumulating about 1600 chunks. That's a lot more than 12 at a time!

**My understanding is that you are proposing strategy 1, combine the single batch.** If that understanding is wrong, I'd appreciate a fuller explanation or code sketch of what you mean.

### Benchmarking the first strategy

I used the following:

```
            if chunks.len() == morsels_per_sink || last_write {
                if maintain_order {
                    chunks.sort_by_key(|chunk| chunk.chunk_index);
                }

                 if !chunks.is_empty() {
                    let mut merged = chunks_to_df_unchecked(std::mem::take(&mut chunks));
                    merged.as_single_chunk();
                    writer._write_batch(&merged).unwrap();
                }

                if last_write {
                    writer._finish().unwrap();
                    return;
                }
            }
```

Now, if run this:

```
$ python ../11699.py 
SINK 1.1473236083984375
COLLECT+WRITE 0.9269287586212158
```

Pretty good! However, if the thread pool is smaller, the code runs twice as slow, because of parquet writing of small chunks:

```
$ POLARS_MAX_THREADS=2 python ../11699.py 
SINK 2.1068201065063477
COLLECT+WRITE 0.9157388210296631
```

In contrast, this PR is unaffected by thread pool size because the rechunking is done by data size, not by number of morsels (== length of `chunks` == thread pool size).

Reply:
> I'm really not getting the impression that repeated vstack_mut leads to quadratic slowdown

No, I don't mean quadratic slowdown. It is a quadratic growing strategy. Which means we might have a few reallocs.

Alright. What I mean is that we need a combination of the strategies. So:

1. Batch chunks across multiple batches. Agree on that
2. Combine them into a single chunk. Also agree on that, but I think we should use `rechunk` (all batches just before flush) instead of `vstack_mut` here.

Reply:
Maybe I'm missing something in the API search, but `rechunk()` seems to be specific to a single `Series`/`DataFrame`? Which would help if the series of `DataChunk` contents was:

```
Series<[2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk]>
Series<[2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk]>
Series<[2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk, 2.5K chunk]>
...
```

But actually the content is:

```
Series<[2.5K chunk]>
Series<[2.5K chunk]>
Series<[2.5K chunk]>
Series<[2.5K chunk]>
...
```

So `rechunk()` on an individual `Series` won't help, you need to combine them somehow, and `vstack_mut` (or perhaps the lower cost `vstack_mut_unchecked`, if it were public) seems to be the only option.

Reply:
Yes. you first combine them into a single `DataFrame` with multiple chunks and then you `rechunk`. Depening on the sink this should be done parallel (`as_single_chunk_par`) or single threaded `as_single_chunk`. Here we should do it single threaded.

So I agree with most of the code here, I just want to use `rechunk` as it allocates properly. If you want I can finish the PR.
So

Reply:
Re-reading all this again, my impression is that you want the _caller_ to decide whether and how to rechunk, instead of, as is currently implemented, `SemicontiguousVstacker` having a fixed policy of calling `as_single_chunk()`.

I tested the code with `as_single_chunk_par`. It appeared to cause a deadlock in the `sink_parquet()` case :grimacing: which I didn't investigate further. It did indeed make the `collect(streaming=True).write_parquet()` version slightly faster.

So I will try implement the above, and if it's still not you meant please do finish the PR, yes.